### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by ogbodo**

### CODE REVIEW

File: StringPrototypeExtension.js [hunk 1]
```diff
@@ -0,0 +1,185 @@
+String.prototype.hasVowels = function() {
```
Comment (mogbeyi-david) [inline StringPrototypeExtension.js]: Please rename this file. Although Pascal casing can be used to name files in a JavaScript project, it is better to use spinal casing to names files at this point. In the event of naming controllers in an API, Pascal casing would suffice as a good naming convention.

File: StringPrototypeExtension.js [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+      }
+      charIndex++;
+      var char = this.charAt(charIndex);
+      if (!/[a-zA-Z]/.test(char)) {
+        alternatingCaseString += char;
+        continue;
+      }
+      alternatingCaseString += this.charAt(charIndex).toUpperCase();
+
+      // if (this.length === charIndex) {
```
Comment (mogbeyi-david) [inline StringPrototypeExtension.js]: Please do away with zombie code

File: StringPrototypeExtension.js [hunk 3]
```diff
@@ -0,0 +1,185 @@
+String.prototype.hasVowels = function() {
+  var vowelRegex = /[aeiou]/gi;
+  var hasVowel = vowelRegex.test(this);
+  return hasVowel;
+};
+
+String.prototype.isQuestion = function() {
+  var isQuestionRegex = /\w+[?]$/;
```
Comment (mogbeyi-david) [inline StringPrototypeExtension.js]: The name of this variable is redundant. The function is already named **isQuestion**. There really is no need to name the regex pattern with the function name in it. Rather put a comment that explains what the regex does, the pattern it matches or something else that is more descriptive of the goal that line is trying to achieve

File: StringPrototypeExtension.js [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+  return isQuestionRegex.test(this);
+};
+
+String.prototype.numberWords = function() {
+  var numberWordRegex = [
+    "zero",
```
Comment (mogbeyi-david) [inline StringPrototypeExtension.js]: Please use single quotes for strings in Javascript

File: StringPrototypeExtension.js [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+    "four",
+    "five",
+    "six",
+    "seven",
+    "eight",
+    "nine"
+  ];
+  var digitsArray = this.match(/\d/g);
+  var numberToWord = "";
+  digitsArray.forEach(digit => {
```
Comment (mogbeyi-david) [inline StringPrototypeExtension.js]: Please stick with ES5 for this project

File: StringPrototypeExtension.js [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+    numberToWord += numberWordRegex[digit].concat(" ");
+  });
+  return numberToWord;
+};
+
+String.prototype.wordsToArray = function() {
+  var wordRegex = /\b[0-9a-zA-Z_\-?]+\b/g;
```
Comment (mogbeyi-david) [inline StringPrototypeExtension.js]: Please add a comment to complex-looking regular expression patterns like this which properly explains what it matches.

File: StringPrototypeExtension.js [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+  var wordArray = this.match(wordRegex);
+
+  return wordArray;
+};
+
+String.prototype.wordCount = function() {
+  return this.wordsToArray().length;
```
Comment (mogbeyi-david) [inline StringPrototypeExtension.js]: Nice! This is a good one-liner, leveraging re-usable code. Good job!

File: StringPrototypeExtension.js [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+};
+
+String.prototype.isDigit = function() {
+  var digitRegex = /\d/g;
+
+  if (!digitRegex.test(this)) {
```
Comment (mogbeyi-david) [inline StringPrototypeExtension.js]: Refactoring this if statement to one line of code would look more elegant. Please remove the braces, put the return statement on the same line as the if statement and add a comment above that line to explain what it is doing.

File: StringPrototypeExtension.js [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+    return false;
+  }
+  return this.match(digitRegex).length === 1 ? true : false;
```
Comment (mogbeyi-david) [inline StringPrototypeExtension.js]: I think this line needs a comment to explain what it accomplishes.

File: StringPrototypeExtension.js [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+};
+
+String.prototype.toCurrency = function() {
+  var regexPattern = /^\d+(\.\d+)?$/g;
```
Comment (mogbeyi-david) [inline StringPrototypeExtension.js]: Please add a comment on this line to explain what the regex matches

File: StringPrototypeExtension.js [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+  if (!regexPattern.test(this)) {
+    return false;
+  }
+  var currency = Number(this)
```
Comment (mogbeyi-david) [inline StringPrototypeExtension.js]: I think this logic of code needs some explaining. Some inbuilt methods are used here and people may not be familiar with these methods or what form of parameters they take. You could add some comments by the side of each logic to explain what is happening at each method

File: StringPrototypeExtension.js [hunk 12]
```diff
@@ -0,0 +1,185 @@
...
+    .toFixed(2)
+    .replace(/\d(?=(\d{3})+\.)/g, "$&,");
```
Comment (mogbeyi-david) [inline StringPrototypeExtension.js]: I think it is better if the regex pattern passed as the first parameter to the replace function is extracted into a well-named variable. This way, the replace function is given more context as to what it does.

File: StringPrototypeExtension.js [hunk 13]
```diff
@@ -0,0 +1,185 @@
...
+
+  if (!digitRegex.test(this)) {
+    return false;
+  }
+  return this.match(digitRegex).length === 1 ? true : false;
+};
+
+String.prototype.toCurrency = function() {
+  var regexPattern = /^\d+(\.\d+)?$/g;
+  if (!regexPattern.test(this)) {
```
Comment (mogbeyi-david) [inline StringPrototypeExtension.js]: Consider reducing this to a single line of code and renaming the variable **regexPattern** or adding a comment to line 54 to explain what the pattern matches

File: StringPrototypeExtension.js [hunk 14]
```diff
@@ -0,0 +1,185 @@
...
+  }
+  var currency = Number(this)
+    .toFixed(2)
+    .replace(/\d(?=(\d{3})+\.)/g, "$&,");
+
+  return currency;
+};
+
+String.prototype.fromCurrency = function() {
+  var regex = /(\d+),(\d+),?(\d+)?(\.)?(\d+)?/;
```
Comment (mogbeyi-david) [inline StringPrototypeExtension.js]: Definitely add a comment here to explain what this pattern matches

File: StringPrototypeExtension.js [hunk 15]
```diff
@@ -0,0 +1,185 @@
...
+  var num = this.replace(regex, "$1$2$3$4$5");
```
Comment (mogbeyi-david) [inline StringPrototypeExtension.js]: Please rename this variable to be more descriptive of the data it contains.

File: StringPrototypeExtension.js [hunk 16]
```diff
@@ -0,0 +1,185 @@
...
+  if (regex.test(this)) {
+    for (var character of this) {
+      var asciiValue = character.charCodeAt();
+      if (asciiValue >= 97 && asciiValue <= 122) {
+        upperCasedString += String.fromCharCode(asciiValue - 32);
+        continue;
+      }
+      upperCasedString += character;
+    }
+  } else {
```
Comment (mogbeyi-david) [inline StringPrototypeExtension.js]: I don't think we need this else statement. You can check if the condition is **NOT** met, return immediately after that and then just write out the logic inside the if statement. This way you maintain a lower number of indentation-levels in your functions and may even be able to get rid of the **"continue"** statement inside the inner if statement.

File: StringPrototypeExtension.js [hunk 17]
```diff
@@ -0,0 +1,185 @@
...
+
+String.prototype.fromCurrency = function() {
+  var regex = /(\d+),(\d+),?(\d+)?(\.)?(\d+)?/;
+  var num = this.replace(regex, "$1$2$3$4$5");
+  return parseFloat(num);
+};
+
+String.prototype.toUpper = function() {
+  var regex = /.+/g;
+  var upperCasedString = "";
```
Comment (mogbeyi-david) [inline StringPrototypeExtension.js]: Please use single quotes for Javascript strings

File: StringPrototypeExtension.js [hunk 18]
```diff
@@ -0,0 +1,185 @@
...
+      lowerCasedString += character;
+    }
+  } else {
+    return false;
+  }
+
+  return lowerCasedString;
+};
+
+String.prototype.ucFirst = function() {
```
Comment (mogbeyi-david) [inline StringPrototypeExtension.js]: I think the name **ucFirst** could be ambiguous. Maybe try a name like **toTitleCase**

File: StringPrototypeExtension.js [hunk 19]
```diff
@@ -0,0 +1,185 @@
...
+  var regex = /\b[a-zA-Z_\-0-9]+[\d\+]?\b/g;
+  var upperCasedFirstString = false;
```
Comment (mogbeyi-david) [inline StringPrototypeExtension.js]: I do not really think this is a good variable name. It does not seem to explain the data it contains in the best way possible.

File: StringPrototypeExtension.js [hunk 20]
```diff
@@ -0,0 +1,185 @@
...
+
+  if (regex.test(this)) {
```
Comment (mogbeyi-david) [inline StringPrototypeExtension.js]: Please check for the NEGATIVE condition first; this way, you would get rid of the else logic

File: StringPrototypeExtension.js [hunk 21]
```diff
@@ -0,0 +1,185 @@
...
+    upperCasedFirstString = "";
+    var arrayOfMatchedString = this.match(regex);
```
Comment (mogbeyi-david) [inline StringPrototypeExtension.js]: I think this variable name should be plural?

File: StringPrototypeExtension.js [hunk 22]
```diff
@@ -0,0 +1,185 @@
...
+
+    arrayOfMatchedString.forEach(word => {
```
Comment (mogbeyi-david) [inline StringPrototypeExtension.js]: Please stick with ES5 functions syntax and do away with ES6 arrow functions for now.

File: __tests__/ExtendedString.test.js [hunk 23]
```diff
@@ -0,0 +1,172 @@
+require("../StringPrototypeExtension");
```
Comment (mogbeyi-david) [inline __tests__/ExtendedString.test.js]: Please rename this test file to be in line with the name of the implementation file in Javascript. Also, rename the test folder from __tests__ to tests.

File: package.json [hunk 24]
```diff
@@ -0,0 +1,23 @@
...
+    "url": "git+https://github.com/ogbodo/StringManipulation.git"
+  },
+  "keywords": [],
+  "author": "",
+  "license": "ISC",
+  "bugs": {
+    "url": "https://github.com/ogbodo/StringManipulation/issues"
+  },
+  "homepage": "https://github.com/ogbodo/StringManipulation#readme",
+  "dependencies": {
```
Comment (mogbeyi-david) [inline package.json]: Please install jest as a devDependency, not a dependency

**Review by mogbeyi-david — CHANGES_REQUESTED** [review]

File: StringPrototypeExtension.js (hunk 7 above)
Comment (ogbodo) [reply StringPrototypeExtension.js]: Thanks

**Review by mogbeyi-david — APPROVED** [review]

**Pull request merged**